File: seed/ruby-sdk-v2/inferred-auth-implicit/lib/seed/internal/inferred_auth_provider.rb
# frozen_string_literal: true

module Seed
  module Internal
    class InferredAuthProvider
      BUFFER_IN_SECONDS = 120 # 2 minutes

      # @param auth_client [untyped]
      # @param options [Hash[String, untyped]]
      #
      # @return [void]
      def initialize(auth_client:, options:)
        @auth_client = auth_client
        @options = options
        @access_token = nil
        @expires_at = nil
      end

      # Returns a cached access token, refreshing if necessary.
      # Refreshes the token if it's nil, or if we're within the buffer period before expiration.
      #
      # @return [String]
      def get_token
        return refresh if @access_token.nil? || token_needs_refresh?

        @access_token
      end

      # Returns the authentication headers to be included in requests.
      #
      # @return [Hash[String, String]]
      def get_auth_headers
        token = get_token
        {
          "Authorization" => "Bearer #{token}"
        }
      end
      # Checks if the token needs to be refreshed.
      # Returns true if the token will expire within the buffer period.
      #
      # @return [Boolean]
      private def token_needs_refresh?
        return true if @expires_at.nil?

        Time.now >= (@expires_at - BUFFER_IN_SECONDS)
      end
      # Refreshes the access token by calling the token endpoint.
      #
      # @return [String]
      private def refresh
        request_params = {
          x_api_key: @options[:x_api_key],
          client_id: @options[:client_id],
          client_secret: @options[:client_secret],
          audience: "https://api.example.com",
          grant_type: "client_credentials",
          scope: @options[:scope]
        }

        token_response = @auth_client.get_token_with_client_credentials(**request_params)

        @access_token = token_response.access_token
        @expires_at = Time.now + token_response.expires_in

        @access_token
      end
    end
  end
end

File: seed/php-sdk/websocket-inferred-auth/src/Core/InferredAuthProvider.php
<?php

namespace Seed\Core;

use Seed\Auth\AuthClient;
use Seed\Auth\Requests\GetTokenRequest;

/**
 * The InferredAuthProvider retrieves an access token from the configured token endpoint.
 * The access token is then used as the bearer token in every authenticated request.
 */
class InferredAuthProvider 
{
    /**
     * @var AuthClient $authClient
     */
    private AuthClient $authClient;

    /**
     * @var array<mixed> $options
     */
    private array $options;

    /**
     * @var ?string $accessToken
     */
    private ?string $accessToken;

    /**
     * @param AuthClient $authClient The client used to retrieve the access token.
     * @param array<mixed> $options The options containing credentials for the token endpoint.
     */
    function __construct(
        AuthClient $authClient,
        array $options,
    )
    {
        $this->authClient = $authClient;
        $this->options = $options;
        $this->accessToken = null;
    }

    /**
     * Returns a cached access token, refreshing if necessary.
     *
     * @return string
     */
    public function getToken(): string {
        if ($this->accessToken !== null){
            return $this->accessToken;
        }
        return $this->refresh();
    }

    /**
     * Returns the authentication headers to be included in requests.
     *
     * @return array<string, string>
     */
    public function getAuthHeaders(): array {
        $token = $this->getToken();
        return [
            'Authorization' => "Bearer " . $token,
        ];
    }

    /**
     * Refreshes the access token by calling the token endpoint.
     *
     * @return string
     */
    private function refresh(): string {
        /** @var array{xApiKey: string, clientId: string, clientSecret: string, audience: 'https://api.example.com', grantType: 'client_credentials', scope?: string|null} $values */
        $values = [
            'xApiKey' => $this->options['xApiKey'] ?? '',
            'clientId' => $this->options['clientId'] ?? '',
            'clientSecret' => $this->options['clientSecret'] ?? '',
            'audience' => 'https://api.example.com',
            'grantType' => 'client_credentials',
            'scope' => $this->options['scope'] ?? null,
        ];

        $request = new GetTokenRequest($values);

        $tokenResponse = $this->authClient->getTokenWithClientCredentials($request);

        $this->accessToken = $tokenResponse->accessToken;

        return $this->accessToken;
    }
}
